### PR TITLE
Glue together for KQ vs K

### PR DESCRIPTION
**Pull request by rajivsambharya**
Up to the point of simulating games for KQ vs K (albeit with essentially random moves).  No training has been performed although the framework should be there.  The main bug I was looking at was the fact that the rollouts always selected the first possible legal action.

Sorry if I changed any of y'alls code.  I also plan on going back and tidying things up tomorrow.

### CODE REVIEW

File: grandmaster.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			chosen_action = np.random.choice(np.arange(actions.size), p=distribution)
+			node = node.outgoing_edges[chosen_action].out_node
+			num_turns += 1
+			if num_turns > 40:
+				winner = 0
+				v = winner*np.ones(num_turns)
+				pi_array = np.array(pi_array)
+				return states, v, pi_array
+		winner = env.outcome(state)
+		v = winner*np.ones(num_turns)
```
Comment (Rulison) [inline grandmaster.py]: Should this be alternating 1s and -1s?

File: mcts.py [hunk 2]
```diff
@@ -41,11 +42,13 @@ def backup(node, value):
         edge.total_action_value += value
         cur_node = edge.in_node
 
-
 def expand_node(node, model, env):
-    action_probs, value = model(node)
-    for i, action_prob in enumerate(action_probs):
-        action = i  # actions are just indexes
+    action_probs, value = model(node.state)
+    legal_actions = env.get_legal_actions(node.state)
```
Comment (thomas-a-neil) [inline mcts.py]: so you want to do this instead of the wrapper in `model.py`?

File: mcts.py [hunk 3]
```diff
@@ -89,6 +92,15 @@ def perform_rollouts(root_node,
 
         n_leaf_expansions -= 1
 
+def get_pi(node, temperature, total_num_actions):
```
Comment (thomas-a-neil) [inline mcts.py]: This seems like it overlaps with the functionality of `get_action_distribution`. Maybe we should combine them into one?

File: grandmaster.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	def __init__(self, model, env, num_games, n_leaf_expansions):
+		self.model = model
+		self.num_games = num_games
+		self.env = env
+		self.start_state = 0
+		self.n_leaf_expansions = n_leaf_expansions
+		self.c = 100
+		self.temperature = 1
+
+	def become_grandmaster():
```
Comment (thomas-a-neil) [inline grandmaster.py]: This should probably take `self` as an argument

File: chess_env.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
+from tests.utils import map_xy_to_square, map_square_to_xy
+import numpy as np
+import chess
+
+class ChessEnv(object):
+	def __init__(self, state_regime, action_regime):
+		self.state_regime = state_regime
+		self.action_regime = action_regime
+		if action_regime == 'KQK_pos_pos_piece':
```
Comment (thomas-a-neil) [inline chess_env.py]: This is probably better managed with a single `True`/`False` flag, and then separate methods for each conversion method.

Comment (thomas-a-neil) [reply chess_env.py]: Or even separate subclasses

File: grandmaster.py (hunk 4 above)
Comment (thomas-a-neil) [reply grandmaster.py]: But it looks like this class isn't tested at all. Did you intend to test it?

File: tests/test_simulate_game.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		action_regime = 'KQK_pos_pos_piece'
+		self.env = ChessEnv(state_regime, action_regime)
+		start_state = np.zeros((8,8,4), dtype=int)
+		start_state[0,2,0] = 1
+		start_state[2,0,1] = 1
+		start_state[3,3,2] = 1
+		self.start_state = start_state
+		sess = tf.Session()
+		self.network = DualNet(sess, state_regime, action_regime)
+		sess.__enter__()
```
Comment (thomas-a-neil) [inline tests/test_simulate_game.py]: Yeah nice I like this in `setUp` actually.

File: tests/test_simulate_game.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+		tf.global_variables_initializer().run()
+
+	def test_simulate_game(self):
```
Comment (thomas-a-neil) [inline tests/test_simulate_game.py]: This test has no assertions, what did you want to test?

File: tests/utils.py [hunk 8]
```diff
@@ -15,7 +14,7 @@ def mock_model_numline(node):
     else:
         value = 0
         action_0_prob = 0.1
-    return (action_0_prob, 1 - action_0_prob), state 
+    return np.array([action_0_prob, 1 - action_0_prob]), state
```
Comment (thomas-a-neil) [inline tests/utils.py]: isn't this supposed to return a value? Why does it return (prob0, prob1), state?

**Review by thomas-a-neil — COMMENTED** [review]
Merp, it changed all my comments (and Jared's) to "outdated" once I merged the formatting changes. You still should be able to see them though.

File: tests/utils.py (hunk 8 above)
Comment (thomas-a-neil) [reply tests/utils.py]: Yes this should be value

File: mcts.py (hunk 3 above)
Comment (thomas-a-neil) [reply mcts.py]: Sounds like they are the same. The difference is that `get_action_distribution` wraps `perform_rollouts` into it.

File: chess_env.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+
+class ChessEnv(object):
+    def __init__(self, state_regime, action_regime):
+        self.state_regime = state_regime
+        self.action_regime = action_regime
+        if action_regime == 'KQK_pos_pos_piece':
+            self.action_dims = (8, 8, 8, 8, 3)
+            self.action_size = 64*64*3
+        elif action_regime == 'KQK_pos_pos':
```
Comment (thomas-a-neil) [inline chess_env.py]: I think these maybe should be two separate classes... this thing isn't really a ChessEnv, it's a simplified env for these two possibilities...

File: game.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+from mcts import exploration_bonus_for_c_puct, perform_rollouts, get_pi
+from tree import Node
+
+
+class Game(object):
+    """docstring for Game"""
+    def self_play(self, model, env, start_state, n_leaf_expansions, c_puct, temperature):
+        node = Node(start_state)
+        pi_array = []
```
Comment (thomas-a-neil) [inline game.py]: Are you using this for debugging? Why are we keeping a vector  of all action distributions?

Comment (thomas-a-neil) [reply game.py]: Nvm I remembered this is used for training 😳

**Pull request merged**